# Обучение моделей P1P2

##### Загрузка данных

In [88]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 50)
pd.set_option('display.precision', 4)

RANDOM_STATE = 1337

In [89]:
from acquiring_xsell.data import load_data
from acquiring_xsell.preprocessing import process_dtypes, fill_missing_values
from acquiring_xsell.features import build_features, make_product_df

df = load_data()
df = process_dtypes(df)
df = fill_missing_values(df)
df = build_features(df)
df = make_product_df(df, "p1p2")

df.head()

,month,inn_status,current_segment_inn,top_mcc_group_inn,who_is_this_first_inn,lifetime_month_streak_inn,real_kam_on_inn,has_termination_start,altpay1_turnover,altpay2_turnover,altpay3_turnover,altpay4_turnover,altpay5_turnover,altpay6_turnover,has_recurrent_payments,cnt_trx,turnover,other_active_product_cnt,is_relevant_mcc_p3,is_relevant_mcc_altpay5,is_relevant_turnover_altpay5,active_mcc_cnt,active_gateway_cnt,avg_check,turnover_ma_2m,turnover_ema_2m,turnover_ma_3m,turnover_ema_3m,turnover_ma_6m,turnover_ema_6m,cnt_trx_ma_2m,cnt_trx_ema_2m,cnt_trx_ma_3m,cnt_trx_ema_3m,cnt_trx_ma_6m,cnt_trx_ema_6m,avg_check_wma_2m,avg_check_wma_3m,avg_check_wma_6m,turnover_cv_2m,turnover_cv_3m,turnover_cv_6m,turnover_change_to_prev_month,cnt_trx_change_to_prev_month,avg_check_change_to_prev_month,turnover_change_to_turnover_ma_2m,turnover_change_to_turnover_ma_3m,turnover_change_to_turnover_ma_6m,turnover_change_to_turnover_ema_2m,turnover_change_to_turnover_ema_3m,turnover_change_to_turnover_ema_6m,cnt_trx_change_to_cnt_trx_ma_2m,cnt_trx_change_to_cnt_trx_ma_3m,cnt_trx_change_to_cnt_trx_ma_6m,cnt_trx_change_to_cnt_trx_ema_2m,cnt_trx_change_to_cnt_trx_ema_3m,cnt_trx_change_to_cnt_trx_ema_6m,avg_check_change_to_avg_check_wma_2m,avg_check_change_to_avg_check_wma_3m,avg_check_change_to_avg_check_wma_6m,is_first_trx_month,is_first_history_month,months_since_last_active_month,months_since_first_trx,has_p3,has_altpay5,rule_flag_p1p2,rule_flag_p3,rule_flag_altpay5,became_active_recently,target
0,2025-04-01,Churn,XXS,Tickets&clubs,referral,0,1,0,0.0,0.0,0.0000,0.0,0.0,0.0,0,0.0000,0.0000,0.0,0,0,0,0,0,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0,0,9,48,0,0,0,0,0,0,0
1,2025-04-01,Churn,XXS,Travel,outcoming,0,1,0,0.0,0.0,0.0000,0.0,0.0,0.0,0,0.0000,0.0000,0.0,0,0,0,0,0,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0,0,8,33,0,0,0,0,0,0,0
2,2025-04-01,Active,XXS,Digital,outcoming,33,0,0,0.0,0.0,0.0000,0.0,0.0,0.0,0,0.0063,0.0009,0.0,0,1,0,1,1,0.0005,0.0044,0.0038,0.0024,0.0035,0.0029,0.0019,0.0103,0.0052,0.0110,0.0144,0.0073,0.0079,0.0016,0.0009,0.0008,0.7867,0.9640,0.6125,-0.7143,-0.3750,-0.5429,-0.5556,-0.4000,-0.5000,-0.3632,-0.4206,-0.4641,-0.2308,0.0000,-0.0385,-0.0833,-0.0774,-0.0577,-0.4222,-0.4000,-0.4800,0,0,0,32,0,0,1,0,0,0,0
3,2025-04-01,Active,M,Digital,outcoming,17,1,0,0.0,0.0,0.0000,0.0,0.0,0.0,0,0.6378,0.6075,0.0,0,1,1,1,1,0.0418,0.6193,0.6119,0.6091,0.6117,0.6369,0.6182,0.6069,0.6321,0.6532,0.6372,0.6257,0.6423,0.0628,0.0382,0.0446,0.0120,0.0634,0.1056,0.0172,0.0270,-0.0096,0.0085,0.0450,-0.0202,0.0072,0.0055,-0.0035,0.0133,0.0703,0.1063,0.0225,0.0418,0.0700,-0.0047,-0.0236,-0.1144,0,0,0,16,0,0,1,0,1,0,0
4,2025-04-01,Active,XS,Charity,referral,36,0,0,0.0,0.0,0.4429,0.0,0.0,0.0,1,0.1768,0.4405,0.0,1,0,1,2,2,0.2072,0.4681,0.4622,0.4383,0.4652,0.4628,0.4501,0.1801,0.1539,0.2115,0.2397,0.1922,0.2072,0.2443,0.1825,0.1970,0.1541,0.1917,0.2269,0.2447,0.0302,0.2081,0.1090,0.2055,0.2004,0.0742,0.1112,0.1537,0.0149,0.0342,0.0119,0.0097,0.0119,0.0124,0.0927,0.1656,0.1863,0,0,0,35,0,0,0,0,0,0,0


## Случайный отбор

In [90]:
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.metrics import format_metrics

train, test = train_test_month_split(df)
K_VALUES = [500, 1000, 1500, 2000]


monthly_ap_train = train.groupby('month')['target'].mean()
monthly_ap_test = test.groupby('month')['target'].mean()

monthly_precision_k_train = {}
monthly_precision_k_test = {}

monthly_recall_k_train = {}
monthly_recall_k_test = {}

monthly_expected_conversions_train = {}
monthly_expected_conversions_test = {}

top_k_train = {}
top_k_test = {}

for k in K_VALUES:

    monthly_precision_k_train[k] = train.groupby('month')['target'].mean()
    monthly_precision_k_test[k] = test.groupby('month')['target'].mean()

    monthly_recall_k_train[k] = k / train.groupby('month').size()
    monthly_recall_k_test[k] = k / test.groupby('month').size()

    monthly_expected_conversions_train[k] = train.groupby('month')['target'].mean() * k
    monthly_expected_conversions_test[k] = test.groupby('month')['target'].mean() * k

    top_k_train[k] = {
        'precision_k': monthly_precision_k_train[k].mean(),
        'recall_k': monthly_recall_k_train[k].mean(),
        'expected_conversions_k': monthly_expected_conversions_train[k].mean()
    }

    top_k_test[k] = {
        'precision_k': monthly_precision_k_test[k].mean(),
        'recall_k': monthly_recall_k_test[k].mean(),
        'expected_conversions_k': monthly_expected_conversions_test[k].mean()
    }


print("Random Selection")

print("\nTrain")
print(f"MAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
display(format_metrics(pd.DataFrame(top_k_train).T))

print("\nTest")
print(f"MAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")
display(format_metrics(pd.DataFrame(top_k_test).T))

Random Selection

Train
MAP: 0.22% ± 0.06%


,precision_k,recall_k,expected_conversions_k
500,0.22%,2.24%,1.12
1000,0.22%,4.48%,2.25
1500,0.22%,6.72%,3.37
2000,0.22%,8.97%,4.50



Test
MAP: 0.25% ± 0.02%


,precision_k,recall_k,expected_conversions_k
500,0.25%,2.17%,1.24
1000,0.25%,4.34%,2.49
1500,0.25%,6.50%,3.73
2000,0.25%,8.67%,4.97


## Отбор по бизнес-правилу

In [91]:
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.metrics import monthly_average_precision_score, all_top_k_scores, format_metrics

train, test = train_test_month_split(df)
y_score_col = 'business_rule_score'

train[y_score_col] = np.where(train['rule_flag_p1p2'] == 1, train['turnover'], 0)
test[y_score_col] = np.where(test['rule_flag_p1p2'] == 1, test['turnover'], 0)


monthly_ap_train = monthly_average_precision_score(train, 'target', y_score_col)
monthly_ap_test = monthly_average_precision_score(test, 'target', y_score_col)

monthly_top_k_train = all_top_k_scores(train, 'target', y_score_col)
monthly_top_k_test = all_top_k_scores(test, 'target', y_score_col)


print("Business-Rule Selection")

print("\nTrain")
print(f"MAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
display(format_metrics(monthly_top_k_train))

print("\nTest")
print(f"MAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")
display(format_metrics(monthly_top_k_test))

Business-Rule Selection

Train
MAP: 0.54% ± 0.13%


,precision_k,recall_k,expected_conversions_k
500,0.67%,6.91%,3.33
1000,0.63%,13.11%,6.25
1500,0.57%,18.15%,8.58
2000,0.52%,21.16%,10.33



Test
MAP: 0.47% ± 0.07%


,precision_k,recall_k,expected_conversions_k
500,0.27%,2.21%,1.33
1000,0.30%,5.36%,3.00
1500,0.36%,9.55%,5.33
2000,0.43%,15.32%,8.67


## Logistic Regression

### Basic

Обучим логистическую регрессию на всех признаках. В рамках Pipeline применим One-Hot Encoding для категориальных признаков и StandardScaler для числовых.

#### Модель без регуляризации

In [92]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_test = test['target']
y_train = train['target']


num_cols = X_train.select_dtypes(include='number').columns
cat_cols = X_train.select_dtypes(include=['object', 'category']).columns

column_transformer = ColumnTransformer(transformers=[
    ('num', StandardScaler(), num_cols),
    ('cat', OneHotEncoder(drop='first', sparse_output=True, dtype=np.int8), cat_cols),
], remainder='passthrough')

lr_pipe = Pipeline([
    ('column_transformer', column_transformer),
    ('lr', LogisticRegression(random_state=RANDOM_STATE))
])

lr_pipe.fit(X_train, y_train)


train_eval = train.copy()
test_eval = test.copy()

train_eval['lr_proba'] = lr_pipe.predict_proba(X_train)[:, 1]
test_eval['lr_proba'] = lr_pipe.predict_proba(X_test)[:, 1]

monthly_ap_train = monthly_average_precision_score(train_eval, 'target', 'lr_proba')
monthly_ap_test = monthly_average_precision_score(test_eval, 'target', 'lr_proba')

print("Logistic Regression")
print(f"Train MAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
print(f"Test  MAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")

Logistic Regression
Train MAP: 3.96% ± 2.40%
Test  MAP: 4.86% ± 1.05%


#### L1

Применим L1-регуляризацию и переберём коэффициенты силы регуляризации.

In [93]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']


num_cols = X_train.select_dtypes(include='number').columns
cat_cols = X_train.select_dtypes(include=['object', 'category']).columns

column_transformer = ColumnTransformer(transformers=[
    ('num', StandardScaler(), num_cols),
    ('cat', OneHotEncoder(drop='first', sparse_output=True, dtype=np.int8), cat_cols),
], remainder='passthrough')


c_values = [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5, 10]
lr_l1_models = {}

print("L1 Logistic Regression")

for c in c_values:
    lr_pipe = Pipeline([
        ('column_transformer', column_transformer),
        ('lr', LogisticRegression(
                        max_iter=100_000,
                        solver="liblinear",
                        l1_ratio=1,
                        C=c,
                        random_state=RANDOM_STATE
                    ))
    ])

    lr_pipe.fit(X_train, y_train)
    lr_l1_models[c] = lr_pipe


    train_eval = train.copy()
    test_eval = test.copy()
    
    train_eval['lr_proba'] = lr_pipe.predict_proba(X_train)[:, 1]
    test_eval['lr_proba'] = lr_pipe.predict_proba(X_test)[:, 1]

    monthly_ap_train = monthly_average_precision_score(train_eval, 'target', 'lr_proba')
    monthly_ap_test = monthly_average_precision_score(test_eval, 'target', 'lr_proba')

    print(f"\nC={c}")
    print(f"Train MAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
    print(f"Test  MAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")

L1 Logistic Regression

C=0.001
Train MAP: 1.48% ± 1.02%
Test  MAP: 2.12% ± 0.82%

C=0.005
Train MAP: 2.49% ± 1.66%
Test  MAP: 3.42% ± 1.34%

C=0.01
Train MAP: 2.97% ± 1.63%
Test  MAP: 4.12% ± 1.90%

C=0.05
Train MAP: 4.29% ± 2.49%
Test  MAP: 4.92% ± 1.73%

C=0.1
Train MAP: 4.21% ± 2.43%
Test  MAP: 4.78% ± 1.46%

C=0.5
Train MAP: 4.40% ± 2.57%
Test  MAP: 4.92% ± 1.51%

C=1
Train MAP: 4.31% ± 2.51%
Test  MAP: 4.84% ± 1.58%

C=5
Train MAP: 4.67% ± 2.88%
Test  MAP: 4.74% ± 1.65%

C=10
Train MAP: 4.70% ± 2.93%
Test  MAP: 4.72% ± 1.62%


#### L2

In [94]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']


num_cols = X_train.select_dtypes(include='number').columns
cat_cols = X_train.select_dtypes(include=['object', 'category']).columns

column_transformer = ColumnTransformer(transformers=[
    ('num', StandardScaler(), num_cols),
    ('cat', OneHotEncoder(drop='first', sparse_output=True, dtype=np.int8), cat_cols),
], remainder='passthrough')


c_values = [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5, 10]
lr_l2_models = {}

print("L2 Logistic Regression")

for c in c_values:
    lr_pipe = Pipeline([
        ('column_transformer', column_transformer),
        ('lr', LogisticRegression(l1_ratio=0, C=c, random_state=RANDOM_STATE))
    ])

    lr_pipe.fit(X_train, y_train)
    lr_l2_models[c] = lr_pipe


    train_eval = train.copy()
    test_eval = test.copy()

    train_eval['lr_proba'] = lr_pipe.predict_proba(X_train)[:, 1]
    test_eval['lr_proba'] = lr_pipe.predict_proba(X_test)[:, 1]

    monthly_ap_train = monthly_average_precision_score(train_eval, 'target', 'lr_proba')
    monthly_ap_test = monthly_average_precision_score(test_eval, 'target', 'lr_proba')

    print(f"\nC={c}")
    print(f"Train MAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
    print(f"Test  MAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")

L2 Logistic Regression

C=0.001
Train MAP: 3.23% ± 1.90%
Test  MAP: 3.78% ± 1.18%

C=0.005
Train MAP: 3.95% ± 2.46%
Test  MAP: 4.69% ± 1.49%

C=0.01
Train MAP: 4.31% ± 2.40%
Test  MAP: 4.81% ± 1.47%

C=0.05
Train MAP: 4.38% ± 2.54%
Test  MAP: 4.61% ± 1.39%

C=0.1
Train MAP: 4.29% ± 2.52%
Test  MAP: 4.57% ± 1.11%

C=0.5
Train MAP: 4.24% ± 2.41%
Test  MAP: 4.71% ± 1.44%

C=1
Train MAP: 3.96% ± 2.40%
Test  MAP: 4.86% ± 1.05%

C=5
Train MAP: 4.15% ± 2.51%
Test  MAP: 4.86% ± 0.98%

C=10
Train MAP: 4.20% ± 2.41%
Test  MAP: 4.62% ± 0.55%


### VIF

#### Отбор признаков

Попробуем итеративно убирать признак с самым большим VIF, пока наибольшее значение VIF в датасете не станет превышать 10.

In [95]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
from acquiring_xsell.preprocessing import train_test_month_split


def get_high_vif_features(data: pd.DataFrame) -> list:

    df = data.copy()
    df_numeric = df.select_dtypes(include='number')

    high_vif_features = []
    max_vif_value = np.inf

    print("High VIF Features\n")

    while max_vif_value > 10:
        curr_df = df_numeric.drop(columns=high_vif_features)

        vif = pd.Series(
            index=curr_df.columns.tolist(),
            data=[variance_inflation_factor(curr_df, i) for i in range(len(curr_df.columns))],
        )

        max_vif_feature = vif.idxmax()
        max_vif_value = vif.max()

        if max_vif_value < 10:
            break
        
        high_vif_features.append(max_vif_feature)

        print(f"{max_vif_feature}, VIF={max_vif_value:.2f}")

    return high_vif_features


train, test = train_test_month_split(df)

HIGH_VIF_FEATURES = get_high_vif_features(train)

High VIF Features

turnover_ema_3m, VIF=1461.66
cnt_trx_change_to_cnt_trx_ema_3m, VIF=1070.07
turnover_ema_2m, VIF=611.61
turnover_change_to_turnover_ema_3m, VIF=518.48
turnover_ma_2m, VIF=475.55
turnover_ma_3m, VIF=303.50
cnt_trx_ema_3m, VIF=266.81
cnt_trx_ema_2m, VIF=184.55
turnover_ma_6m, VIF=175.98
turnover, VIF=142.79
cnt_trx_ma_2m, VIF=122.88
cnt_trx_ema_6m, VIF=108.72
cnt_trx_change_to_cnt_trx_ema_6m, VIF=105.37
avg_check_wma_2m, VIF=71.56
turnover_change_to_turnover_ema_2m, VIF=68.14
cnt_trx_ma_3m, VIF=66.59
avg_check_wma_3m, VIF=42.95
turnover_ema_6m, VIF=41.27
rule_flag_p1p2, VIF=24.94
avg_check_change_to_avg_check_wma_3m, VIF=23.45
turnover_change_to_turnover_ema_6m, VIF=23.32
cnt_trx_change_to_cnt_trx_ma_3m, VIF=19.73
avg_check, VIF=18.81
cnt_trx, VIF=17.66
active_gateway_cnt, VIF=16.49


Обучим логистическую регрессию на признаках без высокого VIF.

#### Модель без регуляризации

In [96]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df.drop(columns=HIGH_VIF_FEATURES))

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']


num_cols = X_train.select_dtypes(include='number').columns
cat_cols = X_train.select_dtypes(include=['object', 'category']).columns

column_transformer = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), num_cols),
        ('cat', OneHotEncoder(drop='first', sparse_output=True, dtype=np.int8), cat_cols),
    ],
    remainder='passthrough'
)

lr_pipe = Pipeline([
    ('column_transformer', column_transformer),
    ('lr', LogisticRegression(random_state=RANDOM_STATE))
])

lr_pipe.fit(X_train, y_train)


train_eval = train.copy()
test_eval = test.copy()

train_eval['lr_proba'] = lr_pipe.predict_proba(X_train)[:, 1]
test_eval['lr_proba'] = lr_pipe.predict_proba(X_test)[:, 1]

monthly_ap_train = monthly_average_precision_score(train_eval, 'target', 'lr_proba')
monthly_ap_test = monthly_average_precision_score(test_eval, 'target', 'lr_proba')

print("Logistic Regression (no High VIF Features)")
print(f"Train MAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
print(f"Test  MAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")

Logistic Regression (no High VIF Features)
Train MAP: 3.77% ± 2.25%
Test  MAP: 4.87% ± 1.29%


#### L1

In [97]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df.drop(columns=HIGH_VIF_FEATURES))

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']


num_cols = X_train.select_dtypes(include='number').columns
cat_cols = X_train.select_dtypes(include=['object', 'category']).columns

column_transformer = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), num_cols),
        ('cat', OneHotEncoder(drop='first', sparse_output=True, dtype=np.int8), cat_cols),
    ],
    remainder='passthrough'
)


c_values = [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5, 10]
lr_l1_models = {}

print("L1 Logistic Regression (no High VIF Features)")

for c in c_values:
    lr_pipe = Pipeline([
        ('column_transformer', column_transformer),
        ('lr', LogisticRegression(
                        max_iter=100_000,
                        solver="liblinear",
                        l1_ratio=1,
                        C=c,
                        random_state=RANDOM_STATE
                    ))
    ])

    lr_pipe.fit(X_train, y_train)
    lr_l1_models[c] = lr_pipe


    train_eval = train.copy()
    test_eval = test.copy()

    train_eval['lr_proba'] = lr_pipe.predict_proba(X_train)[:, 1]
    test_eval['lr_proba'] = lr_pipe.predict_proba(X_test)[:, 1]

    monthly_ap_train = monthly_average_precision_score(train_eval, 'target', 'lr_proba')
    monthly_ap_test = monthly_average_precision_score(test_eval, 'target', 'lr_proba')

    print(f"\nC={c}")
    print(f"Train MAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
    print(f"Test  MAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")

L1 Logistic Regression (no High VIF Features)

C=0.001
Train MAP: 1.48% ± 1.02%
Test  MAP: 2.12% ± 0.82%

C=0.005
Train MAP: 2.49% ± 1.66%
Test  MAP: 3.42% ± 1.34%

C=0.01
Train MAP: 2.76% ± 1.75%
Test  MAP: 3.81% ± 1.50%

C=0.05
Train MAP: 3.98% ± 2.37%
Test  MAP: 4.59% ± 1.62%

C=0.1
Train MAP: 3.91% ± 2.23%
Test  MAP: 4.73% ± 1.70%

C=0.5
Train MAP: 4.01% ± 2.26%
Test  MAP: 4.63% ± 1.55%

C=1
Train MAP: 4.11% ± 2.29%
Test  MAP: 4.58% ± 1.60%

C=5
Train MAP: 4.16% ± 2.41%
Test  MAP: 4.53% ± 1.60%

C=10
Train MAP: 4.17% ± 2.41%
Test  MAP: 4.51% ± 1.62%


#### L2

In [98]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df.drop(columns=HIGH_VIF_FEATURES))

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']


num_cols = X_train.select_dtypes(include='number').columns
cat_cols = X_train.select_dtypes(include=['object', 'category']).columns

column_transformer = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), num_cols),
        ('cat', OneHotEncoder(drop='first', sparse_output=True, dtype=np.int8), cat_cols),
    ],
    remainder='passthrough'
)


c_values = [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5, 10]
lr_l2_models = {}

print("L2 Logistic Regression (no High VIF Features)")

for c in c_values:
    lr_pipe = Pipeline([
        ('column_transformer', column_transformer),
        ('lr', LogisticRegression(l1_ratio=0, C=c, random_state=RANDOM_STATE))
    ])

    lr_pipe.fit(X_train, y_train)
    lr_l2_models[c] = lr_pipe


    train_eval = train.copy()
    test_eval = test.copy()
    
    train_eval['lr_proba'] = lr_pipe.predict_proba(X_train)[:, 1]
    test_eval['lr_proba'] = lr_pipe.predict_proba(X_test)[:, 1]

    monthly_ap_train = monthly_average_precision_score(train_eval, 'target', 'lr_proba')
    monthly_ap_test = monthly_average_precision_score(test_eval, 'target', 'lr_proba')

    print(f"\nC={c}")
    print(f"Train MAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
    print(f"Test  MAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")

L2 Logistic Regression (no High VIF Features)

C=0.001
Train MAP: 3.07% ± 1.64%
Test  MAP: 3.41% ± 1.06%

C=0.005
Train MAP: 3.58% ± 1.97%
Test  MAP: 3.93% ± 1.25%

C=0.01
Train MAP: 3.77% ± 2.06%
Test  MAP: 4.26% ± 1.37%

C=0.05
Train MAP: 4.27% ± 2.16%
Test  MAP: 4.70% ± 1.77%

C=0.1
Train MAP: 3.78% ± 2.23%
Test  MAP: 4.65% ± 1.06%

C=0.5
Train MAP: 3.74% ± 2.22%
Test  MAP: 4.52% ± 0.92%

C=1
Train MAP: 3.77% ± 2.25%
Test  MAP: 4.87% ± 1.29%

C=5
Train MAP: 3.80% ± 2.27%
Test  MAP: 4.89% ± 1.37%

C=10
Train MAP: 3.91% ± 2.27%
Test  MAP: 4.88% ± 1.48%


### Исследование групп признаков

In [99]:
TURNOVER_FEATURES = [
    'turnover',
    'turnover_ma_2m',
    'turnover_ema_2m',
    'turnover_ma_3m',
    'turnover_ema_3m',
    'turnover_ma_6m',
    'turnover_ema_6m',
]

CNT_TRX_FEATURES = [
    'cnt_trx',
    'cnt_trx_ma_2m',
    'cnt_trx_ema_2m',
    'cnt_trx_ma_3m',
    'cnt_trx_ema_3m',
    'cnt_trx_ma_6m',
    'cnt_trx_ema_6m',
]

AVG_CHECK_FEATURES = [
    'avg_check',
    'avg_check_wma_2m',
    'avg_check_wma_3m',
    'avg_check_wma_6m',
]

TURNOVER_CHANGE_FEATURES = [
    'turnover_change_to_prev_month',
    'turnover_change_to_turnover_ma_2m',
    'turnover_change_to_turnover_ma_3m',
    'turnover_change_to_turnover_ma_6m',
    'turnover_change_to_turnover_ema_2m',
    'turnover_change_to_turnover_ema_3m',
    'turnover_change_to_turnover_ema_6m',
]

CNT_TRX_CHANGE_FEATURES = [
    'cnt_trx_change_to_prev_month',
    'cnt_trx_change_to_cnt_trx_ma_2m',
    'cnt_trx_change_to_cnt_trx_ma_3m',
    'cnt_trx_change_to_cnt_trx_ma_6m',
    'cnt_trx_change_to_cnt_trx_ema_2m',
    'cnt_trx_change_to_cnt_trx_ema_3m',
    'cnt_trx_change_to_cnt_trx_ema_6m',
]

AVG_CHECK_CHANGE_FEATURES = [
    'avg_check_change_to_prev_month',
    'avg_check_change_to_avg_check_wma_2m',
    'avg_check_change_to_avg_check_wma_3m',
    'avg_check_change_to_avg_check_wma_6m',
]

TURNOVER_CV_FEATURES = [
    'turnover_cv_2m',
    'turnover_cv_3m',
    'turnover_cv_6m',
]

ACTIVITY_FEATURES = [
    'lifetime_month_streak_inn',
    'months_since_last_active_month',
    'months_since_first_trx',
    'became_active_recently',
]

RULE_FLAG_FEATURES = [
    "rule_flag_p1p2",
    "rule_flag_p3",
    "rule_flag_altpay5",
]

SIMILAR_FEATURE_GROUPS = [
    TURNOVER_FEATURES,
    CNT_TRX_FEATURES,
    AVG_CHECK_FEATURES,
    TURNOVER_CHANGE_FEATURES,
    CNT_TRX_CHANGE_FEATURES,
    AVG_CHECK_CHANGE_FEATURES,
    TURNOVER_CV_FEATURES,
    ACTIVITY_FEATURES,
    RULE_FLAG_FEATURES,
]

#### Влияние групп признаков на MAP

In [100]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from acquiring_xsell.metrics import monthly_average_precision_score

print("Logistic Regression")

for group in SIMILAR_FEATURE_GROUPS:

    train, test = train_test_month_split(df.drop(columns=group))
    
    X_train = train.drop(columns=['target', 'month'])
    X_test = test.drop(columns=['target', 'month'])
    
    y_train = train['target']
    y_test = test['target']
    
    
    num_cols = X_train.select_dtypes(include='number').columns
    cat_cols = X_train.select_dtypes(include=['object', 'category']).columns
    
    column_transformer = ColumnTransformer(
        transformers=[
            ('num', StandardScaler(), num_cols),
            ('cat', OneHotEncoder(drop='first', sparse_output=True, dtype=np.int8), cat_cols),
        ],
        remainder='passthrough'
    )

    lr_pipe = Pipeline([
        ('column_transformer', column_transformer),
        ('lr', LogisticRegression(random_state=RANDOM_STATE))
    ])

    lr_pipe.fit(X_train, y_train)


    train_eval = train.copy()
    test_eval = test.copy()

    train_eval['lr_proba'] = lr_pipe.predict_proba(X_train)[:, 1]
    test_eval['lr_proba'] = lr_pipe.predict_proba(X_test)[:, 1]

    monthly_ap_train = monthly_average_precision_score(train_eval, 'target', 'lr_proba')
    monthly_ap_test = monthly_average_precision_score(test_eval, 'target', 'lr_proba')

    print(f"\nGroup[{group[0]}] Removed")
    print(f"Train MAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
    print(f"Test  MAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")

Logistic Regression

Group[turnover] Removed
Train MAP: 4.11% ± 2.56%
Test  MAP: 4.56% ± 0.60%

Group[cnt_trx] Removed
Train MAP: 4.24% ± 2.44%
Test  MAP: 4.62% ± 1.08%

Group[avg_check] Removed
Train MAP: 3.98% ± 2.42%
Test  MAP: 4.82% ± 1.06%

Group[turnover_change_to_prev_month] Removed
Train MAP: 4.20% ± 2.41%
Test  MAP: 4.92% ± 1.09%

Group[cnt_trx_change_to_prev_month] Removed
Train MAP: 3.99% ± 2.43%
Test  MAP: 4.79% ± 1.15%

Group[avg_check_change_to_prev_month] Removed
Train MAP: 4.35% ± 2.53%
Test  MAP: 4.94% ± 0.79%

Group[turnover_cv_2m] Removed
Train MAP: 4.22% ± 2.57%
Test  MAP: 4.89% ± 0.98%

Group[lifetime_month_streak_inn] Removed
Train MAP: 4.16% ± 2.40%
Test  MAP: 4.75% ± 1.54%

Group[rule_flag_p1p2] Removed
Train MAP: 4.23% ± 2.41%
Test  MAP: 4.71% ± 1.48%


В сравнении с Basic LogReg на всех признаках:
- Качество UP при дропе turnover_change_features, avg_check_change_features
- Больше всего MAP снижается при дропе turnover_features

#### Отбор лучших признаков из групп

In [101]:
train, test = train_test_month_split(df)

HIGH_SIGNAL_FEATURES = []

for group in SIMILAR_FEATURE_GROUPS:

    scores = {}
    for feature in group:
        score = monthly_average_precision_score(train, 'target', feature).mean()
        scores[feature] = score

    best_feature = max(scores, key=scores.get)
    HIGH_SIGNAL_FEATURES.append(best_feature)
    print(f"MAP: {scores[best_feature]:.2%} | Best feature from group: {best_feature}")

LOW_SIGNAL_FEATURES = [feature for group in SIMILAR_FEATURE_GROUPS for feature in group if feature not in HIGH_SIGNAL_FEATURES]

MAP: 0.51% | Best feature from group: turnover_ema_6m
MAP: 0.75% | Best feature from group: cnt_trx_ema_3m
MAP: 0.48% | Best feature from group: avg_check_wma_2m
MAP: 0.57% | Best feature from group: turnover_change_to_prev_month
MAP: 0.64% | Best feature from group: cnt_trx_change_to_prev_month
MAP: 0.39% | Best feature from group: avg_check_change_to_prev_month
MAP: 0.45% | Best feature from group: turnover_cv_2m
MAP: 0.80% | Best feature from group: became_active_recently
MAP: 0.49% | Best feature from group: rule_flag_p1p2


#### Модель без регуляризации

In [102]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df.drop(columns=LOW_SIGNAL_FEATURES))

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']


num_cols = X_train.select_dtypes(include='number').columns
cat_cols = X_train.select_dtypes(include=['object', 'category']).columns

column_transformer = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), num_cols),
        ('cat', OneHotEncoder(drop='first', sparse_output=True, dtype=np.int8), cat_cols),
    ],
    remainder='passthrough'
)

lr_pipe = Pipeline([
    ('column_transformer', column_transformer),
    ('lr', LogisticRegression(random_state=RANDOM_STATE))
])

lr_pipe.fit(X_train, y_train)


train_eval = train.copy()
test_eval = test.copy()

train_eval['lr_proba'] = lr_pipe.predict_proba(X_train)[:, 1]
test_eval['lr_proba'] = lr_pipe.predict_proba(X_test)[:, 1]

monthly_ap_train = monthly_average_precision_score(train_eval, 'target', 'lr_proba')
monthly_ap_test = monthly_average_precision_score(test_eval, 'target', 'lr_proba')

print("Logistic Regression (no Low Signal Features)")
print(f"Train MAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
print(f"Test  MAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")

Logistic Regression (no Low Signal Features)
Train MAP: 4.16% ± 2.49%
Test  MAP: 4.88% ± 1.90%


#### L1

In [103]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df.drop(columns=LOW_SIGNAL_FEATURES))

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']


num_cols = X_train.select_dtypes(include='number').columns
cat_cols = X_train.select_dtypes(include=['object', 'category']).columns

column_transformer = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), num_cols),
        ('cat', OneHotEncoder(drop='first', sparse_output=True, dtype=np.int8), cat_cols),
    ],
    remainder='passthrough'
)


c_values = [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5, 10]
lr_l1_models = {}

print("L1 Logistic Regression (no Low Signal Features)")

for c in c_values:
    lr_pipe = Pipeline([
        ('column_transformer', column_transformer),
        ('lr', LogisticRegression(
                        max_iter=100_000,
                        solver="liblinear",
                        l1_ratio=1,
                        C=c,
                        random_state=RANDOM_STATE
                    ))
    ])

    lr_pipe.fit(X_train, y_train)
    lr_l1_models[c] = lr_pipe


    train_eval = train.copy()
    test_eval = test.copy()
    
    train_eval['lr_proba'] = lr_pipe.predict_proba(X_train)[:, 1]
    test_eval['lr_proba'] = lr_pipe.predict_proba(X_test)[:, 1]

    monthly_ap_train = monthly_average_precision_score(train_eval, 'target', 'lr_proba')
    monthly_ap_test = monthly_average_precision_score(test_eval, 'target', 'lr_proba')

    print(f"\nC={c}")
    print(f"Train MAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
    print(f"Test  MAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")

L1 Logistic Regression (no Low Signal Features)

C=0.001
Train MAP: 1.48% ± 1.02%
Test  MAP: 2.12% ± 0.82%

C=0.005
Train MAP: 2.49% ± 1.66%
Test  MAP: 3.42% ± 1.34%

C=0.01
Train MAP: 2.64% ± 1.71%
Test  MAP: 3.61% ± 1.46%

C=0.05
Train MAP: 4.04% ± 2.52%
Test  MAP: 4.82% ± 1.86%

C=0.1
Train MAP: 4.21% ± 2.41%
Test  MAP: 4.83% ± 1.73%

C=0.5
Train MAP: 4.36% ± 2.50%
Test  MAP: 4.97% ± 1.72%

C=1
Train MAP: 4.28% ± 2.44%
Test  MAP: 4.90% ± 1.66%

C=5
Train MAP: 4.30% ± 2.50%
Test  MAP: 4.89% ± 1.58%

C=10
Train MAP: 4.32% ± 2.54%
Test  MAP: 4.62% ± 1.19%


#### L2

In [104]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df.drop(columns=LOW_SIGNAL_FEATURES))

X_train = train.drop(columns=['target', 'month'])
y_train = train['target']

X_test = test.drop(columns=['target', 'month'])
y_test = test['target']


num_cols = X_train.select_dtypes(include='number').columns
cat_cols = X_train.select_dtypes(include=['object', 'category']).columns

column_transformer = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), num_cols),
        ('cat', OneHotEncoder(drop='first', sparse_output=True, dtype=np.int8), cat_cols),
    ],
    remainder='passthrough'
)


c_values = [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5, 10]
lr_l2_models = {}

print("L2 Logistic Regression (no Low Signal Features)")

for c in c_values:
    lr_pipe = Pipeline([
        ('column_transformer', column_transformer),
        ('lr', LogisticRegression(l1_ratio=0, C=c, random_state=RANDOM_STATE))
    ])

    lr_pipe.fit(X_train, y_train)
    lr_l2_models[c] = lr_pipe


    train_eval = train.copy()
    train_eval['lr_proba'] = lr_pipe.predict_proba(X_train)[:, 1]

    test_eval = test.copy()
    test_eval['lr_proba'] = lr_pipe.predict_proba(X_test)[:, 1]

    monthly_ap_train = monthly_average_precision_score(train_eval, 'target', 'lr_proba')
    monthly_ap_test = monthly_average_precision_score(test_eval, 'target', 'lr_proba')

    print(f"\nC={c}")
    print(f"Train MAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
    print(f"Test  MAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")

L2 Logistic Regression (no Low Signal Features)

C=0.001
Train MAP: 3.15% ± 2.06%
Test  MAP: 3.74% ± 1.31%

C=0.005
Train MAP: 3.93% ± 2.58%
Test  MAP: 4.37% ± 1.61%

C=0.01
Train MAP: 4.07% ± 2.33%
Test  MAP: 4.53% ± 1.83%

C=0.05
Train MAP: 4.72% ± 2.79%
Test  MAP: 4.70% ± 1.89%

C=0.1
Train MAP: 4.41% ± 2.78%
Test  MAP: 5.08% ± 1.83%

C=0.5
Train MAP: 4.43% ± 2.69%
Test  MAP: 4.97% ± 1.75%

C=1
Train MAP: 4.16% ± 2.49%
Test  MAP: 4.88% ± 1.90%

C=5
Train MAP: 4.25% ± 2.60%
Test  MAP: 5.07% ± 2.06%

C=10
Train MAP: 4.53% ± 2.68%
Test  MAP: 5.08% ± 2.04%


### Лучшая модель

Лучшее качество MAP = 5.08% достигнуто на логистической регрессии после отбора признаков с L2-регуляризацией С=0.1, C=10.

In [105]:
from acquiring_xsell.metrics import monthly_average_precision_score, all_top_k_scores, format_metrics

best_c = [0.1, 5, 10]

train, test = train_test_month_split(df.drop(columns=LOW_SIGNAL_FEATURES))

X_train = train.drop(columns=['target', 'month'])
y_train = train['target']

X_test = test.drop(columns=['target', 'month'])
y_test = test['target']


print("Final Logistic Regression Comparison")

for c in best_c:
    lr_pipe = lr_l2_models[c]

    train_eval = train.copy()
    test_eval = test.copy()

    train_eval['lr_proba'] = lr_pipe.predict_proba(X_train)[:, 1]
    test_eval['lr_proba'] = lr_pipe.predict_proba(X_test)[:, 1]

    monthly_ap_train = monthly_average_precision_score(train_eval, 'target', 'lr_proba')
    monthly_ap_test = monthly_average_precision_score(test_eval, 'target', 'lr_proba')

    top_k_train = all_top_k_scores(train_eval, 'target', 'lr_proba')
    top_k_test = all_top_k_scores(test_eval, 'target', 'lr_proba')

    print(f"\nC={c} | L2")
    print(f"Train\nMAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
    display(format_metrics(top_k_train))
    print(f"Test\nMAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")
    display(format_metrics(top_k_test))

Final Logistic Regression Comparison

C=0.1 | L2
Train
MAP: 4.41% ± 2.78%


,precision_k,recall_k,expected_conversions_k
500,3.13%,29.75%,15.67
1000,2.00%,38.51%,20.00
1500,1.48%,43.04%,22.25
2000,1.20%,46.89%,24.08


Test
MAP: 5.08% ± 1.83%


,precision_k,recall_k,expected_conversions_k
500,3.47%,30.22%,17.33
1000,2.37%,40.94%,23.67
1500,1.82%,47.43%,27.33
2000,1.47%,51.06%,29.33



C=5 | L2
Train
MAP: 4.25% ± 2.60%


,precision_k,recall_k,expected_conversions_k
500,3.18%,30.35%,15.92
1000,2.01%,38.47%,20.08
1500,1.51%,44.17%,22.67
2000,1.22%,48.28%,24.50


Test
MAP: 5.07% ± 2.06%


,precision_k,recall_k,expected_conversions_k
500,3.13%,27.36%,15.67
1000,2.20%,38.25%,22.00
1500,1.84%,48.05%,27.67
2000,1.53%,53.41%,30.67



C=10 | L2
Train
MAP: 4.53% ± 2.68%


,precision_k,recall_k,expected_conversions_k
500,3.28%,31.30%,16.42
1000,2.00%,38.46%,20.00
1500,1.51%,44.28%,22.67
2000,1.24%,48.88%,24.83


Test
MAP: 5.08% ± 2.04%


,precision_k,recall_k,expected_conversions_k
500,3.07%,26.73%,15.33
1000,2.07%,35.84%,20.67
1500,1.78%,46.41%,26.67
2000,1.50%,52.19%,30.00


In [106]:
BEST_LOGREG_PIPE = lr_l2_models[0.1]

coefs = pd.Series(
    index=column_transformer.get_feature_names_out(),
    data=BEST_LOGREG_PIPE.named_steps['lr'].coef_[0]
)

print("Top 10 Positive Coefs:")
display(coefs.sort_values(ascending=False).head(10))

print("Top 10 Negative Coefs:")
display(coefs.sort_values(ascending=True).head(10))

Top 10 Positive Coefs:


cat__top_mcc_group_inn_Online&telecom      0.4095
cat__top_mcc_group_inn_Others              0.3698
cat__top_mcc_group_inn_Sharing             0.3604
cat__top_mcc_group_inn_EDU                 0.3573
cat__top_mcc_group_inn_Tickets&clubs       0.3145
num__is_relevant_turnover_altpay5          0.2570
cat__current_segment_inn_XXL               0.2348
cat__top_mcc_group_inn_Electronics/Home    0.1885
cat__top_mcc_group_inn_Fashion/Sport       0.1784
num__is_first_trx_month                    0.1635
dtype: float64

Top 10 Negative Coefs:


cat__top_mcc_group_inn_Charity    -1.4435
num__has_termination_start        -0.9301
cat__top_mcc_group_inn_FoodTech   -0.8889
cat__current_segment_inn_XXS      -0.8867
cat__current_segment_inn_XS       -0.7802
cat__current_segment_inn_S        -0.7743
cat__inn_status_Churn             -0.6720
cat__top_mcc_group_inn_Grocery    -0.6396
cat__top_mcc_group_inn_Churn      -0.5043
cat__inn_status_Reborn            -0.4467
dtype: float64

## LightGBM

### Basic

In [107]:
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.model_selection import MonthSeriesSplit
from lightgbm import LGBMClassifier, early_stopping
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']


cv = MonthSeriesSplit()

fold_scores = []
for train_idx, valid_idx in cv.split(train):

    X_train_cv = X_train.loc[train_idx]
    y_train_cv = y_train.loc[train_idx]

    X_valid_cv = X_train.loc[valid_idx]
    y_valid_cv = y_train.loc[valid_idx]

    lgbm = LGBMClassifier(random_state=RANDOM_STATE, n_estimators=5000, verbose=-1, n_jobs=-1)
    lgbm.fit(
        X_train_cv, y_train_cv,
        eval_set=[(X_valid_cv, y_valid_cv)],
        eval_metric="average_precision",
        callbacks=[early_stopping(50)],
    )

    valid_eval_cv = train.loc[valid_idx].copy()
    valid_eval_cv["lgbm_score"] = lgbm.predict_proba(X_valid_cv)[:, 1]
    
    ap_valid_cv = monthly_average_precision_score(valid_eval_cv, 'target', 'lgbm_score').mean()
    fold_scores.append(ap_valid_cv)


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
X_valid = valid.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_valid = valid['target']
y_test = test['target']


final_model = LGBMClassifier(random_state=RANDOM_STATE, n_estimators=5000, verbose=-1, n_jobs=-1)
final_model.fit(
    X_train, y_train,
    eval_set=[(X_valid, y_valid)],
    eval_metric="average_precision",
    callbacks=[early_stopping(50)],
)


train_eval = train.copy()
test_eval = test.copy()

train_eval["lgbm_score"] = final_model.predict_proba(X_train)[:, 1]
test_eval["lgbm_score"] = final_model.predict_proba(X_test)[:, 1]

monthly_ap_train = monthly_average_precision_score(train_eval, 'target', 'lgbm_score')
monthly_ap_test = monthly_average_precision_score(test_eval, 'target', 'lgbm_score')

print("\nLightGBM")
print(f"Train MAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
print(f"Test  MAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")
print(f"CV    MAP: {np.mean(fold_scores):.2%} ± {np.std(fold_scores):.2%}")

Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[4]	valid_0's average_precision: 0.00717332	valid_0's binary_logloss: 0.0166086
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[2]	valid_0's average_precision: 0.0173827	valid_0's binary_logloss: 0.0193101
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[1]	valid_0's average_precision: 0.0403703	valid_0's binary_logloss: 0.0231949
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[1]	valid_0's average_precision: 0.0403703	valid_0's binary_logloss: 0.0231949

LightGBM
Train MAP: 6.87% ± 3.15%
Test  MAP: 3.91% ± 2.32%
CV    MAP: 2.35% ± 1.49%


### Conservative

In [145]:
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.model_selection import MonthSeriesSplit
from sklearn.model_selection import ParameterSampler
from lightgbm import LGBMClassifier, early_stopping
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']

cv = MonthSeriesSplit()


param_grid = {
    "learning_rate": [0.01, 0.03, 0.05],
    "max_depth": [3, 5, 7],
    "num_leaves": [15, 31, 63],
    "min_child_samples": [50, 100, 200],
    "subsample": [0.7, 0.8, 0.9],
    "colsample_bytree": [0.7, 0.8, 1.0],
}

best_score = 0
best_params = None

for params in ParameterSampler(param_grid, n_iter=30, random_state=RANDOM_STATE):
    
    fold_scores = []
    for train_idx, valid_idx in cv.split(train):

        X_train_cv = X_train.loc[train_idx]
        y_train_cv = y_train.loc[train_idx]

        X_valid_cv = X_train.loc[valid_idx]
        y_valid_cv = y_train.loc[valid_idx]

        model = LGBMClassifier(
            **params,
            random_state=RANDOM_STATE,
            n_estimators=5000,
            n_jobs=-1,
            verbose=-1
        )

        model.fit(
            X_train_cv, y_train_cv,
            eval_set=[(X_valid_cv, y_valid_cv)],
            eval_metric="average_precision",
            callbacks=[early_stopping(50)],
        )

        valid_eval = train.loc[valid_idx].copy()
        valid_eval["score"] = model.predict_proba(X_valid_cv)[:, 1]

        fold_score = monthly_average_precision_score(valid_eval, 'target', 'score').mean()
        fold_scores.append(fold_score)

    mean_score = np.mean(fold_scores)
    std_score = np.std(fold_scores)

    if mean_score > best_score:
        best_score = mean_score
        best_std = std_score
        best_params = params
        print(f"New Best CV MAP: {mean_score:.2%} ± {std_score:.2%}")


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
X_valid = valid.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_valid = valid['target']
y_test = test['target']


lgbm_conservative = LGBMClassifier(
    **best_params,
    random_state=RANDOM_STATE,
    n_estimators=5000,
    n_jobs=-1,
    verbose=-1
)

lgbm_conservative.fit(
    X_train,
    y_train,
    eval_set=[(X_valid, y_valid)],
    eval_metric="average_precision",
    callbacks=[early_stopping(50)]
)


train_eval = train.copy()
valid_eval = valid.copy()
test_eval = test.copy()

train_eval["score"] = lgbm_conservative.predict_proba(X_train)[:, 1]
valid_eval["score"] = lgbm_conservative.predict_proba(X_valid)[:, 1]
test_eval["score"] = lgbm_conservative.predict_proba(X_test)[:, 1]

train_metrics = monthly_average_precision_score(train_eval, 'target', 'score')
valid_metrics = monthly_average_precision_score(valid_eval, 'target', 'score')
test_metrics = monthly_average_precision_score(test_eval, 'target', 'score')


print("\nLightGBM Conservative")
print(f"Best Iteration: {lgbm_conservative.best_iteration_}")
print(f"Best Params: {best_params}")
print(f"Train MAP: {train_metrics.mean():.2%} ± {train_metrics.std():.2%}")
print(f"Valid MAP: {valid_metrics.mean():.2%} ± {valid_metrics.std():.2%}")
print(f"Test  MAP: {test_metrics.mean():.2%} ± {test_metrics.std():.2%}")
print(f"CV    MAP: {best_score:.2%} ± {best_std:.2%}")

Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[149]	valid_0's average_precision: 0.0155119	valid_0's binary_logloss: 0.0112603
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[19]	valid_0's average_precision: 0.032426	valid_0's binary_logloss: 0.0160494
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[36]	valid_0's average_precision: 0.0682055	valid_0's binary_logloss: 0.0173531
New Best CV MAP: 4.33% ± 2.50%
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[12]	valid_0's average_precision: 0.0116382	valid_0's binary_logloss: 0.0118031
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[10]	valid_0's average_precision: 0.0200841	valid_0's binary_logloss: 0.0157197
Training until validation scores don't improve for 50 rounds
Early stopping, best iterati

In [109]:
lgbm_importances = pd.Series(
    index=final_model.feature_names_in_,
    data=final_model.feature_importances_
).sort_values(ascending=False)

print("LightGBM Feature Importances")
print("\nTop 10 Most Important Features:")
print(lgbm_importances.head(10))
print("\nTop 10 Least Important Features:")
print(lgbm_importances.tail(10))

LightGBM Feature Importances

Top 10 Most Important Features:
months_since_first_trx          22
top_mcc_group_inn               14
rule_flag_p1p2                  11
is_relevant_mcc_altpay5         11
altpay3_turnover                10
cnt_trx                          9
is_first_trx_month               8
real_kam_on_inn                  5
avg_check_wma_3m                 5
cnt_trx_change_to_prev_month     4
dtype: int32

Top 10 Least Important Features:
turnover_change_to_turnover_ma_6m    0
turnover_change_to_turnover_ma_3m    0
turnover_change_to_turnover_ma_2m    0
avg_check_change_to_prev_month       0
active_gateway_cnt                   0
turnover_change_to_prev_month        0
turnover_cv_3m                       0
turnover_ema_3m                      0
turnover_ema_6m                      0
inn_status                           0
dtype: int32


### Aggressive

In [110]:
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.model_selection import MonthSeriesSplit
from sklearn.model_selection import ParameterSampler
from lightgbm import LGBMClassifier, early_stopping
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']

cv = MonthSeriesSplit()


param_grid = {
    "learning_rate": [0.1, 0.2, 0.3],
    "max_depth": [6, 8, 10],
    "num_leaves": [63, 127, 255],
    "min_child_samples": [10, 20, 30],
    "subsample": [0.8, 1.0],
    "colsample_bytree": [0.8, 1.0],
}

best_score = 0
best_params = None

for params in ParameterSampler(param_grid, n_iter=30, random_state=RANDOM_STATE):
    
    fold_scores = []
    for train_idx, valid_idx in cv.split(train):

        X_train_cv = X_train.loc[train_idx]
        y_train_cv = y_train.loc[train_idx]

        X_valid_cv = X_train.loc[valid_idx]
        y_valid_cv = y_train.loc[valid_idx]

        model = LGBMClassifier(
            **params,
            random_state=RANDOM_STATE,
            n_estimators=5000,
            n_jobs=-1,
            verbose=-1
        )

        model.fit(
            X_train_cv, y_train_cv,
            eval_set=[(X_valid_cv, y_valid_cv)],
            eval_metric="average_precision",
            callbacks=[early_stopping(50)],
        )

        valid_eval = train.loc[valid_idx].copy()
        valid_eval["score"] = model.predict_proba(X_valid_cv)[:, 1]

        fold_score = monthly_average_precision_score(valid_eval, 'target', 'score').mean()
        fold_scores.append(fold_score)

    mean_score = np.mean(fold_scores)
    std_score = np.std(fold_scores)

    if mean_score > best_score:
        best_score = mean_score
        best_std = std_score
        best_params = params
        print(f"New Best CV MAP: {mean_score:.2%} ± {std_score:.2%}")


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
X_valid = valid.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_valid = valid['target']
y_test = test['target']


final_model = LGBMClassifier(
    **best_params,
    random_state=RANDOM_STATE,
    n_estimators=5000,
    n_jobs=-1,
    verbose=-1
)

final_model.fit(
    X_train,
    y_train,
    eval_set=[(X_valid, y_valid)],
    eval_metric="average_precision",
    callbacks=[early_stopping(50)]
)


train_eval = train.copy()
valid_eval = valid.copy()
test_eval = test.copy()

train_eval["score"] = final_model.predict_proba(X_train)[:, 1]
valid_eval["score"] = final_model.predict_proba(X_valid)[:, 1]
test_eval["score"] = final_model.predict_proba(X_test)[:, 1]

train_metrics = monthly_average_precision_score(train_eval, 'target', 'score')
valid_metrics = monthly_average_precision_score(valid_eval, 'target', 'score')
test_metrics = monthly_average_precision_score(test_eval, 'target', 'score')


print("\nLightGBM Aggressive")
print(f"Best Iteration: {final_model.best_iteration_}")
print(f"Best Params: {best_params}")
print(f"Train MAP: {train_metrics.mean():.2%} ± {train_metrics.std():.2%}")
print(f"Valid MAP: {valid_metrics.mean():.2%} ± {valid_metrics.std():.2%}")
print(f"Test  MAP: {test_metrics.mean():.2%} ± {test_metrics.std():.2%}")
print(f"CV    MAP: {best_score:.2%} ± {best_std:.2%}")

Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[1]	valid_0's average_precision: 0.00686962	valid_0's binary_logloss: 0.0528942
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[1]	valid_0's average_precision: 0.0130942	valid_0's binary_logloss: 0.0450443
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[1]	valid_0's average_precision: 0.0447978	valid_0's binary_logloss: 0.03634
New Best CV MAP: 2.36% ± 1.77%
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[4]	valid_0's average_precision: 0.00530097	valid_0's binary_logloss: 0.017407
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[17]	valid_0's average_precision: 0.00995224	valid_0's binary_logloss: 0.0183746
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration i

### Fine-Tuning

In [ ]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.model_selection import ParameterSampler
from lightgbm import LGBMClassifier, early_stopping
from acquiring_xsell.metrics import monthly_average_precision_score

train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']

cv = MonthSeriesSplit()


param_grid = {
    "learning_rate": [0.003, 0.005, 0.01, 0.02],
    "max_depth": [2, 3, 4],
    "num_leaves": [31, 63, 95, 127],
    "min_child_samples": [150, 200, 300, 400],
    "subsample": [0.7, 0.8, 0.9],
    "colsample_bytree": [0.5, 0.6, 0.7, 0.8],
}

best_score = 0
best_params = None

for params in ParameterSampler(param_grid, n_iter=30, random_state=RANDOM_STATE):
    
    fold_scores = []
    for train_idx, valid_idx in cv.split(train):

        X_train_cv = X_train.loc[train_idx]
        y_train_cv = y_train.loc[train_idx]

        X_valid_cv = X_train.loc[valid_idx]
        y_valid_cv = y_train.loc[valid_idx]

        model = LGBMClassifier(
            **params,
            random_state=RANDOM_STATE,
            n_estimators=5000,
            n_jobs=-1,
            verbose=-1
        )

        model.fit(
            X_train_cv, y_train_cv,
            eval_set=[(X_valid_cv, y_valid_cv)],
            eval_metric="average_precision",
            callbacks=[early_stopping(50)],
        )

        valid_eval = train.loc[valid_idx].copy()
        valid_eval["score"] = model.predict_proba(X_valid_cv)[:, 1]

        fold_score = monthly_average_precision_score(valid_eval, 'target', 'score').mean()
        fold_scores.append(fold_score)

    mean_score = np.mean(fold_scores)
    std_score = np.std(fold_scores)

    if mean_score > best_score:
        best_score = mean_score
        best_std = std_score
        best_params = params
        print(f"New Best CV MAP: {mean_score:.2%} ± {std_score:.2%}")


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
X_valid = valid.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_valid = valid['target']
y_test = test['target']


final_model = LGBMClassifier(
    **best_params,
    random_state=RANDOM_STATE,
    n_estimators=5000,
    n_jobs=-1,
    verbose=-1
)

final_model.fit(
    X_train,
    y_train,
    eval_set=[(X_valid, y_valid)],
    eval_metric="average_precision",
    callbacks=[early_stopping(50)]
)


train_eval = train.copy()
valid_eval = valid.copy()
test_eval = test.copy()

train_eval["score"] = final_model.predict_proba(X_train)[:, 1]
valid_eval["score"] = final_model.predict_proba(X_valid)[:, 1]
test_eval["score"] = final_model.predict_proba(X_test)[:, 1]

train_metrics = monthly_average_precision_score(train_eval, 'target', 'score')
valid_metrics = monthly_average_precision_score(valid_eval, 'target', 'score')
test_metrics = monthly_average_precision_score(test_eval, 'target', 'score')


print("\nLightGBM Fine-tuned")
print(f"Best Iteration: {final_model.best_iteration_}")
print(f"Best Params: {best_params}")
print(f"Train MAP: {train_metrics.mean():.2%} ± {train_metrics.std():.2%}")
print(f"Valid MAP: {valid_metrics.mean():.2%} ± {valid_metrics.std():.2%}")
print(f"Test  MAP: {test_metrics.mean():.2%} ± {test_metrics.std():.2%}")
print(f"CV    MAP: {best_score:.2%} ± {best_std:.2%}")

Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[16]	valid_0's average_precision: 0.0163191	valid_0's binary_logloss: 0.0123343
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[209]	valid_0's average_precision: 0.0281985	valid_0's binary_logloss: 0.0153645
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[48]	valid_0's average_precision: 0.0636907	valid_0's binary_logloss: 0.0178464
New Best CV MAP: 3.86% ± 2.19%
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[5]	valid_0's average_precision: 0.0183974	valid_0's binary_logloss: 0.0125361
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[5]	valid_0's average_precision: 0.0283444	valid_0's binary_logloss: 0.0168046
Training until validation scores don't improve for 50 rounds
Early stopping, best iteratio

### Optuna

In [144]:
import optuna
from lightgbm import LGBMClassifier, early_stopping
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.model_selection import MonthSeriesSplit
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']

cv = MonthSeriesSplit()


def objective(trial):

    params = {
        "learning_rate": trial.suggest_float("learning_rate", 0.005, 0.03, log=True),
        "max_depth": trial.suggest_int("max_depth", 3, 6),
        "num_leaves": trial.suggest_int("num_leaves", 16, 128),
        "min_child_samples": trial.suggest_int("min_child_samples", 100, 500),
        "subsample": trial.suggest_float("subsample", 0.6, 0.95),
        "colsample_bytree": trial.suggest_float("colsample_bytree", 0.6, 1.0),
        "reg_alpha": trial.suggest_float("reg_alpha", 0.0, 5.0), 
        "reg_lambda": trial.suggest_float("reg_lambda", 0.0, 10.0),
    }

    fold_scores = []

    for train_idx, valid_idx in cv.split(train):

        X_train_cv = X_train.loc[train_idx]
        y_train_cv = y_train.loc[train_idx]

        X_valid_cv = X_train.loc[valid_idx]
        y_valid_cv = y_train.loc[valid_idx]

        model = LGBMClassifier(
            **params,
            random_state=RANDOM_STATE,
            n_estimators=5000,
            n_jobs=-1,
            verbose=-1
        )

        model.fit(
            X_train_cv,
            y_train_cv,
            eval_set=[(X_valid_cv, y_valid_cv)],
            eval_metric="average_precision",
            callbacks=[early_stopping(50)]
        )

        valid_eval = train.loc[valid_idx].copy()
        valid_eval["score"] = model.predict_proba(X_valid_cv)[:, 1]

        fold_score = monthly_average_precision_score(valid_eval, 'target', 'score').mean()
        fold_scores.append(fold_score)

    mean_score = np.mean(fold_scores)

    return mean_score


study = optuna.create_study(direction="maximize")
optuna.logging.set_verbosity(optuna.logging.WARNING)

study.optimize(objective, n_trials=50)


best_params = study.best_params
best_score = study.best_value

print(f"Best CV MAP: {best_score:.2%}")
print(f"Best Params: {best_params}")


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
X_valid = valid.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_valid = valid['target']
y_test = test['target']


lgbm_optuna = LGBMClassifier(
    **best_params,
    random_state=RANDOM_STATE,
    n_estimators=5000,
    n_jobs=-1,
    verbose=-1
)
lgbm_optuna.fit(
    X_train,
    y_train,
    eval_set=[(X_valid, y_valid)],
    eval_metric="average_precision",
    callbacks=[early_stopping(50)]
)


train_eval = train.copy()
valid_eval = valid.copy()
test_eval = test.copy()

train_eval["score"] = lgbm_optuna.predict_proba(X_train)[:, 1]
valid_eval["score"] = lgbm_optuna.predict_proba(X_valid)[:, 1]
test_eval["score"] = lgbm_optuna.predict_proba(X_test)[:, 1]


train_metrics = monthly_average_precision_score(train_eval, 'target', 'score')
valid_metrics = monthly_average_precision_score(valid_eval, 'target', 'score')
test_metrics = monthly_average_precision_score(test_eval, 'target', 'score')


print("\nLightGBM Optuna")

print(f"Best Iteration: {lgbm_optuna.best_iteration_}")
print(f"Best Params: {best_params}")
print(f"Train MAP: {train_metrics.mean():.2%} ± {train_metrics.std():.2%}")
print(f"Valid MAP: {valid_metrics.mean():.2%} ± {valid_metrics.std():.2%}")
print(f"Test  MAP: {test_metrics.mean():.2%} ± {test_metrics.std():.2%}")
print(f"CV    MAP: {best_score:.2%}")

Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[260]	valid_0's average_precision: 0.0176953	valid_0's binary_logloss: 0.0113743
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[211]	valid_0's average_precision: 0.0290281	valid_0's binary_logloss: 0.0154128
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[217]	valid_0's average_precision: 0.0630477	valid_0's binary_logloss: 0.016629
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[129]	valid_0's average_precision: 0.017103	valid_0's binary_logloss: 0.0112394
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[133]	valid_0's average_precision: 0.0361081	valid_0's binary_logloss: 0.015207
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[58]	valid_0's average

## CatBoost

#### Basic

In [113]:
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.model_selection import MonthSeriesSplit
from catboost import CatBoostClassifier
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']

cv = MonthSeriesSplit()

cat_cols = X_train.select_dtypes(include=['object', 'category']).columns.tolist()


fold_scores = []
for train_idx, valid_idx in cv.split(train):

    X_train_cv = X_train.loc[train_idx]
    y_train_cv = y_train.loc[train_idx]

    X_valid_cv = X_train.loc[valid_idx]
    y_valid_cv = y_train.loc[valid_idx]

    cb = CatBoostClassifier(
        loss_function="Logloss",
        eval_metric="PRAUC",
        cat_features=cat_cols,
        iterations=5000,
        early_stopping_rounds=50,
        verbose=0,
        random_seed=RANDOM_STATE,
    )

    cb.fit(X_train_cv, y_train_cv, eval_set=(X_valid_cv, y_valid_cv))

    valid_eval_cv = train.loc[valid_idx].copy()
    valid_eval_cv["cb_score"] = cb.predict_proba(X_valid_cv)[:, 1]
    
    ap_valid_cv = monthly_average_precision_score(valid_eval_cv, 'target', 'cb_score').mean()
    fold_scores.append(ap_valid_cv)


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
X_valid = valid.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_valid = valid['target']
y_test = test['target']


final_model = CatBoostClassifier(
    loss_function="Logloss",
    eval_metric="PRAUC",
    cat_features=cat_cols,
    iterations=5000,
    early_stopping_rounds=50,
    verbose=0,
    random_seed=RANDOM_STATE,
)
final_model.fit(X_train, y_train, eval_set=(X_valid, y_valid))


train_eval = train.copy()
valid_eval = valid.copy()
test_eval = test.copy()

train_eval["cb_score"] = final_model.predict_proba(X_train)[:, 1]
valid_eval["cb_score"] = final_model.predict_proba(X_valid)[:, 1]
test_eval["cb_score"] = final_model.predict_proba(X_test)[:, 1]

monthly_ap_train = monthly_average_precision_score(train_eval, 'target', 'cb_score')
monthly_ap_valid = monthly_average_precision_score(valid_eval, 'target', 'cb_score')
monthly_ap_test = monthly_average_precision_score(test_eval, 'target', 'cb_score')

print("CatBoost")
print(f"Best Iteration: {final_model.get_best_iteration()}")
print(f"Train MAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
print(f"Valid MAP: {monthly_ap_valid.mean():.2%} ± {monthly_ap_valid.std():.2%}")
print(f"Test  MAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")
print(f"CV    MAP: {np.mean(fold_scores):.2%} ± {np.std(fold_scores):.2%}")

CatBoost
Best Iteration: 26
Train MAP: 4.54% ± 2.79%
Valid MAP: 7.54% ± 2.27%
Test  MAP: 4.37% ± 1.18%
CV    MAP: 4.10% ± 2.47%


#### Conservative

In [114]:
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.model_selection import MonthSeriesSplit
from sklearn.model_selection import ParameterSampler
from catboost import CatBoostClassifier
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']

cv = MonthSeriesSplit()


param_grid = {
    "learning_rate": [0.01, 0.03, 0.05],
    "depth": [3, 5, 7],
    "l2_leaf_reg": [3, 10, 20],
    "bagging_temperature": [0, 0.5, 1],
    "random_strength": [1, 3, 5],
    "rsm": [0.7, 0.8, 1.0],
}

best_score = 0

for params in ParameterSampler(param_grid, n_iter=30, random_state=RANDOM_STATE):
    
    fold_scores = []
    for train_idx, valid_idx in cv.split(train):

        X_train_cv = X_train.loc[train_idx]
        y_train_cv = y_train.loc[train_idx]

        X_valid_cv = X_train.loc[valid_idx]
        y_valid_cv = y_train.loc[valid_idx]

        cb = CatBoostClassifier(
            **params,
            loss_function="Logloss",
            eval_metric="PRAUC",
            cat_features=cat_cols,
            iterations=5000,
            early_stopping_rounds=50,
            verbose=0,
            random_seed=RANDOM_STATE,
        )

        cb.fit(X_train_cv, y_train_cv, eval_set=(X_valid_cv, y_valid_cv))

        valid_eval_cv = train.loc[valid_idx].copy()
        valid_eval_cv["cb_score"] = cb.predict_proba(X_valid_cv)[:, 1]

        ap_valid_cv = monthly_average_precision_score(valid_eval_cv, 'target', 'cb_score').mean()
        fold_scores.append(ap_valid_cv)

    mean_score = np.mean(fold_scores)
    std_score = np.std(fold_scores)

    if mean_score > best_score:
        best_score = mean_score
        best_std = std_score
        best_params = params
        print(f"New Best CV MAP: {mean_score:.2%} ± {std_score:.2%}")


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
X_valid = valid.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_valid = valid['target']
y_test = test['target']


cb = CatBoostClassifier(
    **best_params,
    loss_function="Logloss",
    eval_metric="PRAUC",
    cat_features=cat_cols,
    iterations=5000,
    early_stopping_rounds=50,
    verbose=0,
    random_seed=RANDOM_STATE,
)

cb.fit(X_train, y_train, eval_set=(X_valid, y_valid))


train_eval = train.copy()
valid_eval = valid.copy()
test_eval = test.copy()

train_eval["score"] = cb.predict_proba(X_train)[:, 1]
valid_eval["score"] = cb.predict_proba(X_valid)[:, 1]
test_eval["score"] = cb.predict_proba(X_test)[:, 1]

train_metrics = monthly_average_precision_score(train_eval, 'target', 'score')
valid_metrics = monthly_average_precision_score(valid_eval, 'target', 'score')
test_metrics = monthly_average_precision_score(test_eval, 'target', 'score')


print("CatBoost Conservative")
print(f"Best Iteration: {cb.best_iteration_}")
print(f"Best Params: {best_params}")
print(f"Train MAP: {train_metrics.mean():.2%} ± {train_metrics.std():.2%}")
print(f"Valid MAP: {valid_metrics.mean():.2%} ± {valid_metrics.std():.2%}")
print(f"Test  MAP: {test_metrics.mean():.2%} ± {test_metrics.std():.2%}")
print(f"CV    MAP: {best_score:.2%} ± {best_std:.2%}")

New Best CV MAP: 3.95% ± 1.99%
New Best CV MAP: 4.71% ± 2.89%
New Best CV MAP: 4.75% ± 2.31%
CatBoost Conservative
Best Iteration: 105
Best Params: {'rsm': 0.8, 'random_strength': 5, 'learning_rate': 0.05, 'l2_leaf_reg': 3, 'depth': 7, 'bagging_temperature': 0.5}
Train MAP: 8.00% ± 4.39%
Valid MAP: 7.66% ± 1.35%
Test  MAP: 4.20% ± 0.71%
CV    MAP: 4.75% ± 2.31%


#### Aggressive

In [115]:
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.model_selection import MonthSeriesSplit
from sklearn.model_selection import ParameterSampler
from catboost import CatBoostClassifier
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']

cv = MonthSeriesSplit()


param_grid = {
    "learning_rate": [0.1, 0.2, 0.3],
    "depth": [6, 8, 10],
    "l2_leaf_reg": [1, 3, 5],
    "bagging_temperature": [0, 0.3],
    "random_strength": [0, 1],
    "rsm": [0.8, 1.0],
}

best_score = 0

for params in ParameterSampler(param_grid, n_iter=30, random_state=RANDOM_STATE):
    
    fold_scores = []
    for train_idx, valid_idx in cv.split(train):

        X_train_cv = X_train.loc[train_idx]
        y_train_cv = y_train.loc[train_idx]

        X_valid_cv = X_train.loc[valid_idx]
        y_valid_cv = y_train.loc[valid_idx]

        cb = CatBoostClassifier(
            **params,
            loss_function="Logloss",
            eval_metric="PRAUC",
            cat_features=cat_cols,
            iterations=5000,
            early_stopping_rounds=50,
            verbose=0,
            random_seed=RANDOM_STATE,
        )

        cb.fit(X_train_cv, y_train_cv, eval_set=(X_valid_cv, y_valid_cv))

        valid_eval_cv = train.loc[valid_idx].copy()
        valid_eval_cv["cb_score"] = cb.predict_proba(X_valid_cv)[:, 1]

        ap_valid_cv = monthly_average_precision_score(valid_eval_cv, 'target', 'cb_score').mean()
        fold_scores.append(ap_valid_cv)

    mean_score = np.mean(fold_scores)
    std_score = np.std(fold_scores)

    if mean_score > best_score:
        best_score = mean_score
        best_std = std_score
        best_params = params
        print(f"New Best CV MAP: {mean_score:.2%} ± {std_score:.2%}")


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
X_valid = valid.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_valid = valid['target']
y_test = test['target']


cb = CatBoostClassifier(
    **best_params,
    loss_function="Logloss",
    eval_metric="PRAUC",
    cat_features=cat_cols,
    iterations=5000,
    early_stopping_rounds=50,
    verbose=0,
    random_seed=RANDOM_STATE,
)

cb.fit(X_train, y_train, eval_set=(X_valid, y_valid))


train_eval = train.copy()
valid_eval = valid.copy()
test_eval = test.copy()

train_eval["score"] = cb.predict_proba(X_train)[:, 1]
valid_eval["score"] = cb.predict_proba(X_valid)[:, 1]
test_eval["score"] = cb.predict_proba(X_test)[:, 1]

train_metrics = monthly_average_precision_score(train_eval, 'target', 'score')
valid_metrics = monthly_average_precision_score(valid_eval, 'target', 'score')
test_metrics = monthly_average_precision_score(test_eval, 'target', 'score')


print("CatBoost Aggressive")
print(f"Best Iteration: {cb.best_iteration_}")
print(f"Best Params: {best_params}")
print(f"Train MAP: {train_metrics.mean():.2%} ± {train_metrics.std():.2%}")
print(f"Valid MAP: {valid_metrics.mean():.2%} ± {valid_metrics.std():.2%}")
print(f"Test  MAP: {test_metrics.mean():.2%} ± {test_metrics.std():.2%}")
print(f"CV    MAP: {best_score:.2%} ± {best_std:.2%}")

New Best CV MAP: 4.58% ± 2.33%
New Best CV MAP: 5.12% ± 3.63%
CatBoost Aggressive
Best Iteration: 35
Best Params: {'rsm': 0.8, 'random_strength': 0, 'learning_rate': 0.1, 'l2_leaf_reg': 3, 'depth': 8, 'bagging_temperature': 0}
Train MAP: 11.91% ± 4.48%
Valid MAP: 10.17% ± 2.00%
Test  MAP: 4.55% ± 0.91%
CV    MAP: 5.12% ± 3.63%


#### Fine-Tuning

In [117]:
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.model_selection import MonthSeriesSplit
from sklearn.model_selection import ParameterSampler
from catboost import CatBoostClassifier
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']

cv = MonthSeriesSplit()


param_grid = {
    "learning_rate": [0.03, 0.05, 0.07, 0.1],
    "depth": [6, 7, 8],
    "l2_leaf_reg": [3, 5, 7, 10],
    "bagging_temperature": [0, 0.3, 0.5],
    "random_strength": [0, 1, 3],
    "rsm": [0.7, 0.8, 0.9],
}

best_score = 0

for params in ParameterSampler(param_grid, n_iter=30, random_state=RANDOM_STATE):
    
    fold_scores = []
    for train_idx, valid_idx in cv.split(train):

        X_train_cv = X_train.loc[train_idx]
        y_train_cv = y_train.loc[train_idx]

        X_valid_cv = X_train.loc[valid_idx]
        y_valid_cv = y_train.loc[valid_idx]

        cb = CatBoostClassifier(
            **params,
            loss_function="Logloss",
            eval_metric="PRAUC",
            cat_features=cat_cols,
            iterations=5000,
            early_stopping_rounds=50,
            verbose=0,
            random_seed=RANDOM_STATE,
        )

        cb.fit(X_train_cv, y_train_cv, eval_set=(X_valid_cv, y_valid_cv))

        valid_eval_cv = train.loc[valid_idx].copy()
        valid_eval_cv["cb_score"] = cb.predict_proba(X_valid_cv)[:, 1]

        ap_valid_cv = monthly_average_precision_score(valid_eval_cv, 'target', 'cb_score').mean()
        fold_scores.append(ap_valid_cv)

    mean_score = np.mean(fold_scores)
    std_score = np.std(fold_scores)

    if mean_score > best_score:
        best_score = mean_score
        best_std = std_score
        best_params = params
        print(f"New Best CV MAP: {mean_score:.2%} ± {std_score:.2%}")


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
X_valid = valid.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_valid = valid['target']
y_test = test['target']


cb = CatBoostClassifier(
    **best_params,
    loss_function="Logloss",
    eval_metric="PRAUC",
    cat_features=cat_cols,
    iterations=5000,
    early_stopping_rounds=50,
    verbose=0,
    random_seed=RANDOM_STATE,
)

cb.fit(X_train, y_train, eval_set=(X_valid, y_valid))


train_eval = train.copy()
valid_eval = valid.copy()
test_eval = test.copy()

train_eval["score"] = cb.predict_proba(X_train)[:, 1]
valid_eval["score"] = cb.predict_proba(X_valid)[:, 1]
test_eval["score"] = cb.predict_proba(X_test)[:, 1]

train_metrics = monthly_average_precision_score(train_eval, 'target', 'score')
valid_metrics = monthly_average_precision_score(valid_eval, 'target', 'score')
test_metrics = monthly_average_precision_score(test_eval, 'target', 'score')


print("CatBoost Fine-tuned")
print(f"Best Iteration: {cb.best_iteration_}")
print(f"Best Params: {best_params}")
print(f"Train MAP: {train_metrics.mean():.2%} ± {train_metrics.std():.2%}")
print(f"Valid MAP: {valid_metrics.mean():.2%} ± {valid_metrics.std():.2%}")
print(f"Test  MAP: {test_metrics.mean():.2%} ± {test_metrics.std():.2%}")
print(f"CV    MAP: {best_score:.2%} ± {best_std:.2%}")

New Best CV MAP: 4.54% ± 2.30%
New Best CV MAP: 4.60% ± 1.80%
CatBoost Fine-tuned
Best Iteration: 44
Best Params: {'rsm': 0.9, 'random_strength': 3, 'learning_rate': 0.1, 'l2_leaf_reg': 10, 'depth': 8, 'bagging_temperature': 0.5}
Train MAP: 8.50% ± 4.70%
Valid MAP: 6.54% ± 0.37%
Test  MAP: 3.80% ± 0.45%
CV    MAP: 4.60% ± 1.80%


#### Optuna

In [128]:
import optuna
from catboost import CatBoostClassifier
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.model_selection import MonthSeriesSplit
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']

cat_cols = X_train.select_dtypes(include=['object', 'category']).columns.tolist()

cv = MonthSeriesSplit()


def objective(trial):

    params = {
        "learning_rate": trial.suggest_float("learning_rate", 0.005, 0.1, log=True),
        "depth": trial.suggest_int("depth", 3, 10),
        "l2_leaf_reg": trial.suggest_float("l2_leaf_reg", 1, 20, log=True),
        "bagging_temperature": trial.suggest_float("bagging_temperature", 0.0, 2.0),
        "random_strength": trial.suggest_float("random_strength", 0.0, 5.0),
        "rsm": trial.suggest_float("rsm", 0.5, 1.0),
        "border_count": trial.suggest_int("border_count", 32, 255),
    }

    fold_scores = []

    for train_idx, valid_idx in cv.split(train):

        X_train_cv = X_train.loc[train_idx]
        y_train_cv = y_train.loc[train_idx]

        X_valid_cv = X_train.loc[valid_idx]
        y_valid_cv = y_train.loc[valid_idx]

        model = CatBoostClassifier(
            **params,
            loss_function="Logloss",
            eval_metric="PRAUC",
            iterations=5000,
            early_stopping_rounds=50,
            verbose=0,
            random_seed=RANDOM_STATE,
            cat_features=cat_cols
        )

        model.fit(
            X_train_cv,
            y_train_cv,
            eval_set=(X_valid_cv, y_valid_cv)
        )

        valid_eval = train.loc[valid_idx].copy()
        valid_eval["score"] = model.predict_proba(X_valid_cv)[:, 1]

        fold_score = monthly_average_precision_score(
            valid_eval, 'target', 'score'
        ).mean()

        fold_scores.append(fold_score)

    return np.mean(fold_scores)


optuna.logging.set_verbosity(optuna.logging.WARNING)

study = optuna.create_study(direction="maximize")
study.optimize(objective, n_trials=50)

best_params = study.best_params
best_score = study.best_value

print(f"Best CV MAP: {best_score:.2%}")
print(f"Best Params: {best_params}")


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
X_valid = valid.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_valid = valid['target']
y_test = test['target']

final_model = CatBoostClassifier(
    **best_params,
    loss_function="Logloss",
    eval_metric="PRAUC",
    iterations=5000,
    early_stopping_rounds=50,
    verbose=0,
    random_seed=RANDOM_STATE,
    cat_features=cat_cols
)

final_model.fit(
    X_train,
    y_train,
    eval_set=(X_valid, y_valid)
)


train_eval = train.copy()
valid_eval = valid.copy()
test_eval = test.copy()

train_eval["score"] = final_model.predict_proba(X_train)[:, 1]
valid_eval["score"] = final_model.predict_proba(X_valid)[:, 1]
test_eval["score"] = final_model.predict_proba(X_test)[:, 1]


train_metrics = monthly_average_precision_score(train_eval, 'target', 'score')
valid_metrics = monthly_average_precision_score(valid_eval, 'target', 'score')
test_metrics = monthly_average_precision_score(test_eval, 'target', 'score')

print("\nCatBoost Optuna")

print(f"Best Iteration: {final_model.get_best_iteration()}")
print(f"Best Params: {best_params}")
print(f"Train MAP: {train_metrics.mean():.2%} ± {train_metrics.std():.2%}")
print(f"Valid MAP: {valid_metrics.mean():.2%} ± {valid_metrics.std():.2%}")
print(f"Test  MAP: {test_metrics.mean():.2%} ± {test_metrics.std():.2%}")
print(f"CV    MAP: {best_score:.2%}")

Best CV MAP: 5.21%
Best Params: {'learning_rate': 0.03827443398073261, 'depth': 9, 'l2_leaf_reg': 1.9708101938759761, 'bagging_temperature': 1.2183502080061959, 'random_strength': 2.99589707247177, 'rsm': 0.5024733270884685, 'border_count': 255}

CatBoost Optuna
Best Iteration: 54
Best Params: {'learning_rate': 0.03827443398073261, 'depth': 9, 'l2_leaf_reg': 1.9708101938759761, 'bagging_temperature': 1.2183502080061959, 'random_strength': 2.99589707247177, 'rsm': 0.5024733270884685, 'border_count': 255}
Train MAP: 6.88% ± 2.81%
Valid MAP: 8.88% ± 4.09%
Test  MAP: 4.52% ± 0.67%
CV    MAP: 5.21%


## Лучшая модель

In [152]:
from acquiring_xsell.preprocessing import train_test_month_split
from acquiring_xsell.model_selection import MonthSeriesSplit
from sklearn.model_selection import ParameterSampler
from lightgbm import LGBMClassifier, early_stopping
from acquiring_xsell.metrics import monthly_average_precision_score


train, test = train_test_month_split(df)

X_train = train.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_test = test['target']

cv = MonthSeriesSplit()


param_grid = {
    "learning_rate": [0.01, 0.03, 0.05],
    "max_depth": [3, 5, 7],
    "num_leaves": [15, 31, 63],
    "min_child_samples": [50, 100, 200],
    "subsample": [0.7, 0.8, 0.9],
    "colsample_bytree": [0.7, 0.8, 1.0],
}

best_score = 0
best_params = None

for params in ParameterSampler(param_grid, n_iter=30, random_state=RANDOM_STATE):
    
    fold_scores = []
    for train_idx, valid_idx in cv.split(train):

        X_train_cv = X_train.loc[train_idx]
        y_train_cv = y_train.loc[train_idx]

        X_valid_cv = X_train.loc[valid_idx]
        y_valid_cv = y_train.loc[valid_idx]

        model = LGBMClassifier(
            **params,
            random_state=RANDOM_STATE,
            n_estimators=5000,
            n_jobs=-1,
            verbose=-1
        )

        model.fit(
            X_train_cv, y_train_cv,
            eval_set=[(X_valid_cv, y_valid_cv)],
            eval_metric="average_precision",
            callbacks=[early_stopping(50)],
        )

        valid_eval = train.loc[valid_idx].copy()
        valid_eval["score"] = model.predict_proba(X_valid_cv)[:, 1]

        fold_score = monthly_average_precision_score(valid_eval, 'target', 'score').mean()
        fold_scores.append(fold_score)

    mean_score = np.mean(fold_scores)
    std_score = np.std(fold_scores)

    if mean_score > best_score:
        best_score = mean_score
        best_std = std_score
        best_params = params
        print(f"New Best CV MAP: {mean_score:.2%} ± {std_score:.2%}")


train, valid, test = train_test_month_split(df, get_valid=True)

X_train = train.drop(columns=['target', 'month'])
X_valid = valid.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_valid = valid['target']
y_test = test['target']


lgbm_conservative = LGBMClassifier(
    **best_params,
    random_state=RANDOM_STATE,
    n_estimators=5000,
    n_jobs=-1,
    verbose=-1
)

lgbm_conservative.fit(
    X_train,
    y_train,
    eval_set=[(X_valid, y_valid)],
    eval_metric="average_precision",
    callbacks=[early_stopping(50)]
)


train_eval = train.copy()
valid_eval = valid.copy()
test_eval = test.copy()

train_eval["score"] = lgbm_conservative.predict_proba(X_train)[:, 1]
valid_eval["score"] = lgbm_conservative.predict_proba(X_valid)[:, 1]
test_eval["score"] = lgbm_conservative.predict_proba(X_test)[:, 1]

train_metrics = monthly_average_precision_score(train_eval, 'target', 'score')
valid_metrics = monthly_average_precision_score(valid_eval, 'target', 'score')
test_metrics = monthly_average_precision_score(test_eval, 'target', 'score')


print("\nLightGBM Conservative")
print(f"Best Iteration: {lgbm_conservative.best_iteration_}")
print(f"Best Params: {best_params}")
print(f"Train MAP: {train_metrics.mean():.2%} ± {train_metrics.std():.2%}")
print(f"Valid MAP: {valid_metrics.mean():.2%} ± {valid_metrics.std():.2%}")
print(f"Test  MAP: {test_metrics.mean():.2%} ± {test_metrics.std():.2%}")
print(f"CV    MAP: {best_score:.2%} ± {best_std:.2%}")

Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[149]	valid_0's average_precision: 0.0155119	valid_0's binary_logloss: 0.0112603
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[19]	valid_0's average_precision: 0.032426	valid_0's binary_logloss: 0.0160494
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[36]	valid_0's average_precision: 0.0682055	valid_0's binary_logloss: 0.0173531
New Best CV MAP: 4.33% ± 2.50%
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[12]	valid_0's average_precision: 0.0116382	valid_0's binary_logloss: 0.0118031
Training until validation scores don't improve for 50 rounds
Early stopping, best iteration is:
[10]	valid_0's average_precision: 0.0200841	valid_0's binary_logloss: 0.0157197
Training until validation scores don't improve for 50 rounds
Early stopping, best iterati

In [153]:
best_model = lgbm_conservative

### Калибровка Платта

In [154]:
from acquiring_xsell.preprocessing import train_test_month_split
from sklearn.linear_model import LogisticRegression
from acquiring_xsell.metrics import monthly_average_precision_score, all_top_k_scores, format_metrics

train, valid, test = train_test_month_split(df, get_valid=True)


X_train = train.drop(columns=['target', 'month'])
X_valid = valid.drop(columns=['target', 'month'])
X_test = test.drop(columns=['target', 'month'])

y_train = train['target']
y_valid = valid['target']
y_test = test['target']


train_scores = best_model.predict_proba(X_train)[:, 1]
valid_scores = best_model.predict_proba(X_valid)[:, 1]
test_scores = best_model.predict_proba(X_test)[:, 1]


platt_model = LogisticRegression(random_state=RANDOM_STATE)
platt_model.fit(valid_scores.reshape(-1, 1), y_valid)


train_eval = train.copy()
valid_eval = valid.copy()
test_eval = test.copy()

train_eval['platt_score'] = platt_model.predict_proba(train_scores.reshape(-1, 1))[:, 1]
valid_eval['platt_score'] = platt_model.predict_proba(valid_scores.reshape(-1, 1))[:, 1]
test_eval['platt_score'] = platt_model.predict_proba(test_scores.reshape(-1, 1))[:, 1]

monthly_ap_train = monthly_average_precision_score(train_eval, 'target', 'platt_score')
monthly_ap_valid = monthly_average_precision_score(valid_eval, 'target', 'platt_score')
monthly_ap_test = monthly_average_precision_score(test_eval, 'target', 'platt_score')

top_k_train = all_top_k_scores(train_eval, 'target', 'platt_score')
top_k_valid = all_top_k_scores(valid_eval, 'target', 'platt_score')
top_k_test = all_top_k_scores(test_eval, 'target', 'platt_score')


print(f"Train\nMAP: {monthly_ap_train.mean():.2%} ± {monthly_ap_train.std():.2%}")
display(format_metrics(top_k_train))
print(f"Valid\nMAP: {monthly_ap_valid.mean():.2%} ± {monthly_ap_valid.std():.2%}")
display(format_metrics(top_k_valid))
print(f"Test\nMAP: {monthly_ap_test.mean():.2%} ± {monthly_ap_test.std():.2%}")
display(format_metrics(top_k_test))

Train
MAP: 4.80% ± 2.63%


,precision_k,recall_k,expected_conversions_k
500,2.98%,32.48%,14.89
1000,1.76%,38.50%,17.56
1500,1.33%,43.74%,19.89
2000,1.12%,49.26%,22.44


Valid
MAP: 7.79% ± 1.14%


,precision_k,recall_k,expected_conversions_k
500,4.93%,37.88%,24.67
1000,2.83%,43.11%,28.33
1500,2.07%,47.78%,31.00
2000,1.68%,52.02%,33.67


Test
MAP: 5.10% ± 0.04%


,precision_k,recall_k,expected_conversions_k
500,4.20%,36.27%,21.00
1000,2.43%,42.28%,24.33
1500,1.87%,48.76%,28.00
2000,1.55%,53.97%,31.00


In [155]:
from joblib import dump
from pathlib import Path

dump(best_model, Path("..", "models", "p1p2.joblib"))
dump(platt_model, Path("..", "models", "p1p2_platt.joblib"))

['../models/p1p2_platt.joblib']